### PR TITLE
New Auth Method Added.

### PR DESCRIPTION
**Pull request by kenglou**
Added the following Authentication Provider:
1. Game Center.
2. Google Play Game Services.
3. Xiao Mi

### CODE REVIEW

File: src/Adapters/Auth/index.js [hunk 1]
```diff
@@ -1,10 +1,18 @@
-import loadAdapter from '../AdapterLoader';
+'use strict';
+
+var _AdapterLoader = require('../AdapterLoader');
+
+var _AdapterLoader2 = _interopRequireDefault(_AdapterLoader);
+
+function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
```
Comment (dplewis) [inline src/Adapters/Auth/index.js]: Looks like you copy compiled babel. Doesn't look like something humans write.

File: src/Adapters/Auth/index.js [hunk 2]
```diff
@@ -41,36 +51,40 @@ const providers = {
...
-
-function authDataValidator(adapter, appIds, options) {
-  return function(authData) {
-    return adapter.validateAuthData(authData, options).then(() => {
-      if (appIds) {
+  weibo,
+  xiaomi
+};
+
+function authDataValidator(adapter, appIds, options)
```
Comment (dplewis) [inline src/Adapters/Auth/index.js]: Sticking to the current coding style / ES6 would be ideal.

File: src/Adapters/Auth/index.js [hunk 3]
```diff
@@ -41,36 +51,40 @@ const providers = {
...
+  };
 }
 
 function loadAuthAdapter(provider, authOptions) {
   const defaultAdapter = providers[provider];
   const adapter = Object.assign({}, defaultAdapter);
   const providerOptions = authOptions[provider];
 
-  if (!defaultAdapter && !providerOptions) {
+  if (!defaultAdapter && !providerOptions)
```
Comment (dplewis) [inline src/Adapters/Auth/index.js]: Its recommended to always use brackets for if / else statements. Without them they can lead to bugs. You might want to go over your code again for consistency

File: src/Adapters/Auth/index.js [hunk 4]
```diff
@@ -41,36 +51,40 @@ const providers = {
...
     return;
-  }
 
   const appIds = providerOptions ? providerOptions.appIds : undefined;
 
   // Try the configuration methods
   if (providerOptions) {
-    const optionalAdapter = loadAdapter(providerOptions, undefined, providerOptions);
+    const optionalAdapter = (0, _AdapterLoader2.default)(providerOptions, undefined, providerOptions);
```
Comment (dplewis) [inline src/Adapters/Auth/index.js]: Babel

**Comment by kenglou** [comment]
already changed accordingly. Please let me know if there is anything else.

File: src/Adapters/Auth/index.js [hunk 5]
```diff
@@ -41,18 +45,22 @@ const providers = {
...
+    return adapter.validateAuthData(authData, options).then(() =>
+    {
+      if (appIds)
+      {
         return adapter.validateAppId(appIds, authData, options);
       }
       return Promise.resolve();
     });
-  }
+  };
```
Comment (montymxb) [inline src/Adapters/Auth/index.js]: Can you revert these style changes? 

Generally you should avoid unnecessary stylistic changes when adding a feature unless it's part of the change itself (i.e. specifically a stylistic change PR). In this case it's superfluous and adds to what we have to review. It'll make it _much_ easier for us to go through and review your changes if they're only functional changes.

File: src/Adapters/Auth/index.js [hunk 6]
```diff
@@ -70,7 +78,7 @@ function loadAuthAdapter(provider, authOptions) {
   if (providerOptions) {
     const optionalAdapter = loadAdapter(providerOptions, undefined, providerOptions);
     if (optionalAdapter) {
-      ['validateAuthData', 'validateAppId'].forEach((key) => {
+      ['validateAuthData', 'validateAppId'].forEach(key => {
```
Comment (montymxb) [inline src/Adapters/Auth/index.js]: Same here. This can and should be suggested separately.

File: src/Adapters/Auth/index.js [hunk 7]
```diff
@@ -82,18 +90,26 @@ function loadAuthAdapter(provider, authOptions) {
     return;
   }
 
-  return {adapter, appIds, providerOptions};
+  return { adapter, appIds, providerOptions };
```
Comment (montymxb) [inline src/Adapters/Auth/index.js]: and here

File: src/Adapters/Auth/index.js [hunk 8]
```diff
@@ -82,18 +90,26 @@ function loadAuthAdapter(provider, authOptions) {
...
-  const setEnableAnonymousUsers = function(enable) {
+  const setEnableAnonymousUsers = function (enable)
+  {
     _enableAnonymousUsers = enable;
-  }
+  };
   // To handle the test cases on configuration
-  const getValidatorForProvider = function(provider) {
+  const getValidatorForProvider = function (provider)
+  {
```
Comment (montymxb) [inline src/Adapters/Auth/index.js]: and above here

File: src/Adapters/Auth/index.js [hunk 9]
```diff
@@ -104,12 +120,12 @@ module.exports = function(authOptions = {}, enableAnonymousUsers = true) {
...
-  }
+  };
 
   return Object.freeze({
     getValidatorForProvider,
     setEnableAnonymousUsers
-  })
-}
+  });
+};
```
Comment (montymxb) [inline src/Adapters/Auth/index.js]: Not needed for this PR, let's kept it focused on what you're actually introducing.

File: src/Adapters/Auth/gcenter.js [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+// Helper functions for accessing the google API.
+//var Parse = require('parse/node').Parse;
+var crypto = require('crypto');
+var request = require('request');
+var url = require('url');
+
+// Returns a promise that fulfills if this user id is valid.
+function validateAuthData(authData)
+{
```
Comment (montymxb) [inline src/Adapters/Auth/gcenter.js]: bring this opening brace back up to the function declaration, keeping in style with the rest of everything.

File: src/Adapters/Auth/gpgames.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+// Helper functions for accessing the google API.
+var https = require('https');
+var Parse = require('parse/node').Parse;
+var request = require('request');
+
+// Returns a promise that fulfills if this user id is valid.
+function validateAuthData(authData, authOptions)
+{
```
Comment (montymxb) [inline src/Adapters/Auth/gpgames.js]: again, keep the curly braces up on the same line. It's inconsistent throughout this file as well if you can amend that.

File: src/Adapters/Auth/xiaomi.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+// Helper functions for accessing the Xiao Mi Graph API.
+var https = require('https');
+var Parse = require('parse/node').Parse;
+var crypto = require('crypto');
+
+const PRODUCTION_URL = "https://cn-api.unity.com";
+const DEBUG_URL = " https://cn-api-debug.unity.com";
```
Comment (montymxb) [inline src/Adapters/Auth/xiaomi.js]: This url doesn't seem right, xiaomi is unity? Can you clarify this 3rd party provider and the urls?

**Pull request closed without merging**